### PR TITLE
Re-work sighash type conversion methods

### PR DESCRIPTION
**Pull request by tcharding**
This PR has evolved into a full blown clean up of the conversion methods for all three sighash types based on the discussion below.

Everything is split up into very small patches to aid review (and bikeshedding, this PR is almost totally just naming things).

EDIT: I'm not convinced this should be an RC-fix, the changes are too wide spread now. It started as just a single method rename. Leaving label as is for others to consider.

### CODE REVIEW

**Comment by sanket1729** [comment]
This is also good, I also wouldn't mind nuking the inner method. Your preference.

**Review by sanket1729 — APPROVED** [review]
utACK 5efa0067344cbc9b03cc6c5f70729254e82a04c5

**Comment by tcharding** [comment]
Let's nuke it :)

**Review by sanket1729 — DISMISSED** [review]
ACK bec898e055d3b4024aa5221396dd5153b426557c

File: src/util/psbt/map/input.rs
Comment (Kixunil) [inline src/util/psbt/map/input.rs:187]: Did you mean to write `to_u32`? You wrote it in the PR comment that way and I think it should be `to_u32` because it doesn't return a reference and performs copy.

While we're renaming things shouldn't this be something like `to_consensus_u32`? (I'm not sure which representation it is.) And maybe fix the doc claiming it returns "byte" while you're touching that part.

File: src/util/psbt/serialize.rs
Comment (Kixunil) [inline src/util/psbt/serialize.rs:175]: I prefer to call methods so that if the representation changes it's easier to keep it in sync. (Not that it'd be likely here but may be worth reinforcing good habits.)

**Review by Kixunil — COMMENTED** [review]
I think this deserves at least a comment but wouldn't block it.

File: src/util/psbt/serialize.rs
Comment (tcharding) [reply src/util/psbt/serialize.rs:175]: oh yeah? I was unsure of this and explicitly thought about it so I'm glad you mentioned it. I thought because we were _inside_ the module to use the field not the public function. You're suggesting to _always_ use the function if one exists. Lesson learned, thanks.

File: src/util/psbt/map/input.rs
Comment (tcharding) [reply src/util/psbt/map/input.rs]: I went with `to_u32` because that is Rust convention. This PR highlights that we have a bunch of these as/to/into things wrong. I'll do the others as a separate PR.

Comment (tcharding) [reply src/util/psbt/map/input.rs]: Follow up PR #798

Comment (Kixunil) [reply src/util/psbt/map/input.rs]: Why not name the representation? We were using names that clearly state the representation elsewhere and I think it's very useful. I wouldn't even say it breaks Rust convention. (If it does then this would be the exceptional case when I disagree with the convention.)

**Review by apoelstra — DISMISSED** [review]
ACK f627f67626f3640af01c368094043a83e7d95c5c

File: src/util/psbt/map/input.rs
Comment (tcharding) [reply src/util/psbt/map/input.rs]: My bad for not responding, I did have a reason. A `EcdsaSigHashType` could have been constructed with either `from_u32_consensus` _or_ `from_u32_standard`. When one calls `to_u32` on that object we do not know which it is standard or consensus so I figured `to_u32` was the best name. Thoughts?

Comment (Kixunil) [reply src/util/psbt/map/input.rs]: Well, it's consensus *encoding* (we did not map it to our internal representation). `from_u32_standard` is also consensus encoding, it's just more restrictive.

Because our type does funny things to non-standard values, I believe it should be called `to_u32_standard`. Also standard implies consensus.

Comment (tcharding) [reply src/util/psbt/map/input.rs]: I'm hoping this PR can go into the 0.28 release. Can we leave the improvement of `to_u32` -> `to_u32_standard` until next release so that we can change `EcdsaSigHashType::as_u32` with it?

Comment (tcharding) [reply src/util/psbt/map/input.rs]: I've added a link to this thread to my todo list so it doesn't get forgotten.

Comment (Kixunil) [reply src/util/psbt/map/input.rs]: I don't like changing the same thing twice, especially knowing we will. Can we just change it once and well?

Comment (tcharding) [reply src/util/psbt/map/input.rs]: No worries, PR updated and force pushed.

**Comment by tcharding** [comment]
@apoelstra please re-ack, if you agree, PR now uses `to_u32_standard`. Thanks.

**Review by Kixunil — APPROVED** [review]
ACK bb17582862146bdb83611e54c917e845f7a4241c

**Review by apoelstra — DISMISSED** [review]
ACK bb17582862146bdb83611e54c917e845f7a4241c

I think the new name is a good idea, given our current crappy handling of non-standard flags.

**Comment by tcharding** [comment]
ouch, I was on the wrong branch and rebased on master. Sorry fellas, no changes made. Please re-ack and give me a virtual slap.

**Review by Kixunil — DISMISSED** [review]
ACK 9de614e6d9914194020861765c3739ce8e1cfcbc

**Review by apoelstra — APPROVED** [review]
ACK 9de614e6d9914194020861765c3739ce8e1cfcbc

**Comment by sanket1729** [comment]
Thanks for the bump, I have been exclusively looking at taproot release PRs. Feel free to tag me for review. 

I will look at your amazing places, but so far I was only tunnel visioning on 0.28  milestone. I missed the RC fix label.

**Comment by sanket1729** [comment]
> I think the new name is a good idea, given our current crappy handling of non-standard flags.

This is not true. It is possible to have non-standard values if they are explicitly constructed from serialize::Deserialize with non-standard u32. I would rather it be called `to_u32`. 

The purpose of `PsbtSigHashType` is to capture non-standard sighash types, but users would need their own parsing code to deal with it. All high level APIs like `ecdsa_hash_ty` and `schnorr_hash_ty` will fail and the user would need to parse the sighash byte separately and deal with it. No information should be lost in ser/de using psbt serialize methods.

**Comment by apoelstra** [comment]
Ok, given @sanket1729's comments I rescind my ACK

**Comment by Kixunil** [comment]
> It is possible to have non-standard values if they are explicitly constructed from serialize::Deserialize with non-standard u32.

This sounds massively broken. Either we should enforce it's standard in all constructors or in neither. I believe we agreed standard is better?

**Comment by apoelstra** [comment]
My recollection was that we agreed "only support standard for now but we should revisit this when we have more room to make breaking API changes". I don't remember whether "only support standard" extended to preventing de/serialization of nonstandard flags.

I agree that the current behaviour is "massively broken" but I think it represents an awkward (and temporary!) compromise.

**Comment by Kixunil** [comment]
> I don't remember whether "only support standard" extended to preventing de/serialization of nonstandard

Me neither but I strongly believe it should. The next release is massively-breaking anyway, so I'm not convinced being breaking is a good argument against it.

**Comment by sanket1729** [comment]
I believe `PsbtSigHashType` addresses all concerns for non-standard sighash types too. The methods to and from `EcdsaSigHashType` and `SchnorrSigHashType` return error on invalid/non-standard sighash types. The ser/de explicitly maintains the inner value. You can do something as follows to deal with non-standard things

```rust
if let Err(NonStandardSigHashType) = psbt.input[idx].sighash_ty.ecdsa_sighash_ty() {
	let sighash_u32 = psbt.input[idx].sighash_ty.as_u32();
}
```
 Does this address all concerns?

**Comment by sanket1729** [comment]
There is however no direct constructor for making `PsbtSigHashType` type. It has to be created by `serialize::Deserialize`. I believe that adding a  `PsbtSigHashType::from_non_standard_u32` or `PsbtSigHashType::from_u32_raw` should be enough.

**Comment by Kixunil** [comment]
Ah, all these types are starting to confuse me (or rather myself not remembering all of them). Let's check if I got this right:

* We have `EcdsaSigHashType`, `SchnorrSigHashType`, `PsbtSigHashType`
* `PsbtSigHashType` doesn't enforce anything - so it's consensus compatible. Therefore it should make sense to have methods `to_consensus` and `from_consensus`
* The other two sighash types enforce standardness among other things, they should have methods `to_standard` and `from_standard`
* `PsbtSIgHashType` can be infallibly created from other two and fallibly converted to them.

Is this correct?

BTW I noticed `PsbtSigHashType` is missing link in docs.rs which seems to be caused by us forgetting to reexport it in a public module?

**Comment by tcharding** [comment]
>  BTW I noticed PsbtSigHashType is missing link in docs.rs which seems to be caused by us forgetting to reexport it in a public module?

I have not attended to this one, FTR the type is re-exported in the `psbt` module which is public. If you have it handy @Kixunil can you share the command again that builds the docs as they are built for docs.rs please? You told it to me before and I defined a shell alias for it but forgot to sync my shell rc files before going traveling :(

**Comment by tcharding** [comment]
> Is this correct?

I went to read the code to see if you were correct and all these patches fell out of it :) I think you were/are correct, except for I added the integer type in the method names e.g. `from_u8_standard`.

**Comment by Kixunil** [comment]
I'm referring to this: https://docs.rs/bitcoin/0.28.0-rc.1/bitcoin/util/psbt/struct.Input.html#structfield.sighash_type

The command is `RUSTDOCFLAGS="--cfg docsrs" cargo +nightly doc --features std,secp-recovery,base64,rand,use-serde,bitcoinconsensus --open`

Integer type in method should be fine (although it has Hungarian vibes).

**Comment by tcharding** [comment]
> Integer type in method should be fine (although it has Hungarian vibes).

I agree, we had other methods already there with the integer in the name so I went with that. I can change them all to remove them but this is already getting way too involved for a rc fix. Thoughts?

**Comment by tcharding** [comment]
I'll fix the 1.29 CI fails tomorrow.

**Comment by Kixunil** [comment]
> already getting way too involved for a rc fix

Yeah, I think so. It'll suck a bit if we change it then but we have these instances all over the crate, so I don't think it matters much.

**Comment by tcharding** [comment]
Since I commented above that I believe this should _not_ be an rc-fix I elected to remove all the `_u*_` integer types from the methods added/changed by this PR as suggested.

File: src/util/sighash.rs
Comment (sanket1729) [inline src/util/sighash.rs:153]: There is no standardness for `SchnorrSignaturehashType`. It is either valid or invalid. Ecdsa one has a standardness rule. This is also reflected in the error variant names. Perhaps this should be renamed back to `from_u8`?

**Comment by sanket1729** [comment]
Mostly ACK, modulo the schnorr signature changes. As there is no notion of standardness validity for those, I think we should call them `from_u8`. (or any other better name)

I get that adding return types in signatures might not be ideal, but I think we should make this one exception for ecdsa sighash types because they are sometimes serialized as 1 byte when attached to signatures, and serialized as 4 bytes while computing sighash.

File: src/util/sighash.rs
Comment (Kixunil) [reply src/util/sighash.rs:153]: I view it as standardness here being equivalent with consensus validity. :) Does `from_consensus` seem less confusing to you?

**Comment by Kixunil** [comment]
> they are sometimes serialized as 1 byte when attached to signatures, and serialized as 4 bytes while computing sighash

That's an interesting argument though I think it's still better to refer to these cases by their *purpose*. Something like `serialize_for_signature()` and `serialize_for_sighash()`, please bikeshed.

**Comment by apoelstra** [comment]
@Kixunil I find `u8` and `u32` much clearer as to what they do and where they'd be used. `for_signature` and `for_sighash` I would need to read the function bodies of to check that they actually do what I need them to do.

**Comment by Kixunil** [comment]
Hmm, maybe have both purpose and type?

**Comment by apoelstra** [comment]
I think the type implies the purpose. If you are using these methods, you're writing consensus-related signature code and need to know all of the rules, and your logic would be obfuscated by having long names for everything.

**Comment by Kixunil** [comment]
I'm not really sure. I often read docs (or code) of ambiguous things like this. "What does this value mean?" Even if I think it does the expected thing.
That's also why I always document non-obvious impls and such.

**Comment by tcharding** [comment]
This discussion is getting super confusing, attempting to find clarity/consensus.

> I get that adding return types in signatures might not be ideal, but I think we should make this one exception for ecdsa sighash types because they are sometimes serialized as 1 byte when attached to signatures, and serialized as 4 bytes while computing sighash.

@sanket1729 is this comment typed correctly? Specifically is the 'ecdsa' correct or did you mean 'schnorr'?

We don't currently provide a way to get `EcdsaSigHashType` as a `u8`. (For `SchnorrSigHashType` we return a `u8` which can be cast safely to a `u32`).

Furthermore:

- We have not reached consensus on the naming of the `EcdsaSigHashType` conversion method that returns a `u32`, currently it is `as_u32`, in this PR it is changed to `to_standard`.
- We have _loose_ consensus that to/from_u32 is best for `PsbtSigHashType`.

If I've missed anyone's view point the fault is my own, please re-state them :)

**Comment by sanket1729** [comment]
There are two different issues: 
Issue 1: serialized types dependent on modes of operation

Ecdsa signature hash as u32 in signature hash caculation methods for `segwit_signature_hash` and `legacy_signature_hash`. 
https://github.com/rust-bitcoin/rust-bitcoin/blob/2c1077e6817cbe1f04d6668dda56b62d920e0c92/src/util/sighash.rs#L567

Ecdsa signature hash as u8 in signatures when they are serialized as witnesses in transactions. 
https://github.com/rust-bitcoin/rust-bitcoin/blob/2c1077e6817cbe1f04d6668dda56b62d920e0c92/src/util/ecdsa.rs#L62

But for schnorr signatures, it is always serialized as u8 (both in sighash calculation and signature serialization)

Issue 2: Standardness
EcdsaSigHashTypes are raw `u8` in consensus, but only a few values listed in `EcdsaSigHashType` are allowed as per standardness rules. When used in sighash calculation it is serialized as u32. Other raw `u8` values not listed in `EcdsaSigHashType` are consensus valid, but not standardness valid.

There is no notion for standardness validity for `SchnorrSignatureType`. They are either a valid `SchnorrSignatureHashType` or just invalid.

Based on the above, In my opinion:

> We have loose consensus that to/from_u32 is best for PsbtSigHashType.

 Agreed with this.

> We have not reached consensus on the naming of the EcdsaSigHashType conversion method that returns a u32, currently it is as_u32, in this PR it is changed to to_standard.

I/ (slightly)prefer the names `as_u32` party because 1) I am used to it. 2) And `to_standard` would be confusing to me because the type can also support standard things. I would not mind changing it to `as_sighash_msg_u32` (feel free to suggest names here), but I don't feel too strongly about this API would be rarely used in practice. Most people should use `legacy_signature_hash` API or `segwit_signature_hash` API that will use the correct serialization. People who would use this are expected to read the BIP for exact ser/de and would pay careful attention to the name API to use.

But, I do feel strongly about the `u8` ser/de because that is what almost everyone will be directly using. I think our current API does a good job at cleanly erroring on non-standard values. and only serializing to standard values. 

As for the schnorr API naming, I prefer if those don't involve standard/consensus names in them. We designed the new soft forks clearly so that we can have cleaner implementations without worrying about these things. I prefer not confusing users with consensus/standard names when it is really not necessary.

**Comment by tcharding** [comment]
Thanks  for your input @sanket1729.

> I/ (slightly)prefer the names as_u32 party because 

Did you mean `to_u32` not `as_u32`?

> 1) I am used to it. 2) And to_standard would be confusing to me because the type can also support standard things. 

I've used `to_u32`since it is objectively more correct (idiomatic) than `as_32`. I have added to the commit message a note that we have not managed to reach consensus on this naming.

> But, I do feel strongly about the u8 ser/de because that is what almost everyone will be directly using. I think our current API does a good job at cleanly erroring on non-standard values. and only serializing to standard values.

I'm not sure if the current state of the PR is in line with this comment or violates it?

> As for the schnorr API naming, I prefer if those don't involve standard/consensus names in them. We designed the new soft forks clearly so that we can have cleaner implementations without worrying about these things. I prefer not confusing users with consensus/standard names when it is really not necessary.

I've dropped the `from_u32` -> `from_consensus` change since we do not have consensus on it either.

**Comment by tcharding** [comment]
With the above discussion in mind the changes in this PR represent an improvement to the code but _not_ a perfect solution. I've tried to distill out the things that are good to merge given that this is still marked as rc-fix. Further improvements and reaching consensus is left for another day.

**Comment by tcharding** [comment]
Changes in rebase:
- Fixed bug in docs, indentation was incorrect
- Improved docs, added backticks
- Re-based on master

**Review by apoelstra — DISMISSED** [review]
ACK 618ce282dde2e0c57a14452913c3842531024159

**Comment by tcharding** [comment]
This is the last RC-fix labelled open PR, is this all that's holding up the Taproot release or have we some other work to do?

**Comment by tcharding** [comment]
Are you ok with this one @Kixunil, I know a bunch of your wishes were not implemented in its current state. I did my best to get some resolution though.

File: src/util/psbt/map/input.rs
Comment (dr-orlovsky) [inline src/util/psbt/map/input.rs:190]: Nit: please not block because of this one

```suggestion
    /// ([`EcdsaSigHashType`], [`SchnorrSigHashType`] respectively).
```

File: src/util/sighash.rs
Comment (dr-orlovsky) [inline src/util/sighash.rs:134]: Nit: I think this may be useful outside of the scope of just this library (for instance for some wallet applications)

File: src/util/sighash.rs
Comment (dr-orlovsky) [inline src/util/sighash.rs:143]: Not sure that `Reserved` will not be changed in the future and some of its variants may support `ANYONECANPAY`. May be the function signature should be updated to return Optional as a second parameter?

**Review by dr-orlovsky — COMMENTED** [review]
Two nits (which should be ignored if this PR will get more ACKs) and one question

File: src/util/psbt/map/input.rs
Comment (tcharding) [reply src/util/psbt/map/input.rs:190]: Good catch, fixed as suggested.

File: src/util/sighash.rs
Comment (tcharding) [reply src/util/sighash.rs:134]: Fair point but unrelated to this PR so leaving as is. If we want this to be public we should do it as an RC fix though, I would not be confident to write the supporting PR message championing the idea other than to copy what you have said verbatim. Can I leave it with you?

File: src/util/sighash.rs
Comment (tcharding) [reply src/util/sighash.rs:143]: As above, I think this is unrelated to this PR since these lines show up as a refactor (move code block) only. And as above I would not be able to explain the change, can I leave this one with you also?

**Comment by tcharding** [comment]
Change in force-push: Used links as suggested by @dr-orlovsky

**Comment by tcharding** [comment]
> @tcharding I'd like to also update our version of rust-secp in this RC.

Did PR for you: https://github.com/rust-bitcoin/rust-bitcoin/pull/875

**Comment by tcharding** [comment]
Thanks for the review @dr-orlovsky

**Review by dr-orlovsky — APPROVED** [review]
ACK 3e316721dda7415844184a19456e431395445e2f

**Review by apoelstra — DISMISSED** [review]
ACK 3e316721dda7415844184a19456e431395445e2f

**Review by sanket1729 — APPROVED** [review]
ACK 3e316721dda7415844184a19456e431395445e2f. Sorry for the delayed response, I am really happy with the state of this PR now. Thanks for keeping with the reviews.

**Comment by apoelstra** [comment]
Sorry, needs rebase because of #806. But yes, @Kixunil we need your approval.

**Comment by tcharding** [comment]
No changes in force push, rebase only.

**Review by sanket1729 — APPROVED** [review]
utACK 05d4dc988e17623a2aa9471e76c8b28d9f16d9a8

**Review by dr-orlovsky — DISMISSED** [review]
ACK 05d4dc988e17623a2aa9471e76c8b28d9f16d9a8

Not merging to wait for @Kixunil approval since he had comments before

**Comment by tcharding** [comment]
Changes in force push:
- I added back ticks to a `u32` in the docs change in `06e7b09 Rename as_u32 -> to_u32`.
- rebased

While checking over the diff before force pushing I noticed that this PR adds a `From<&str>` method that immediately calls `to_string()` on the argument. Do you want me to change this to a `From<String>` method @Kixunil and then call `to_string` at the call site? (Noticed while reviewing: https://github.com/rust-bitcoin/rust-bitcoin/pull/898)

**Comment by tcharding** [comment]
This PR had a bunch of unrelated changes to the `SchnorrSigHashType`, these were conflicting with work being done by @dr-orlovsky in https://github.com/rust-bitcoin/rust-bitcoin/pull/898. I've pulled all those changes out and put them in a [separate PR](https://github.com/rust-bitcoin/rust-bitcoin/pull/903) so we can more easily handle the conflicts.

**Review by sanket1729 — APPROVED** [review]
ACK 51ac67356d6b581a99f4775546962a964c18b77b. In the interest of moving things forward, I think we should merge this after it meets the requirements(2 ACKs and no Nacks/comments for other reviewers). 

We can address @Kixunil new comments in follow-up PR after the next RC or if required in the next breaking release.

**Comment by dr-orlovsky** [comment]
ACK 51ac67356d6b581a99f4775546962a964c18b77b

Not merging yet to give @apoelstra and @Kixunil another day to comment.

**Comment by dr-orlovsky** [comment]
ouh, this now requires rebase :( :( :(

**Comment by sanket1729** [comment]
I think we should focus on getting this in before any other merge now :)

**Comment by tcharding** [comment]
Force push is rebase only, no other changes.

**Comment by tcharding** [comment]
@Kixunil where have you gone man, are you ok with this one? Its not perfect but is it acceptable?

**Review by sanket1729 — APPROVED** [review]
utACK 8e2422f92bcc8f7b8264d262c166826afe8a93ca.

**Review by dr-orlovsky — APPROVED** [review]
ACK 8e2422f92bcc8f7b8264d262c166826afe8a93ca

**Comment by dr-orlovsky** [comment]
Merging now, if there will be major comments from other reviewers will do a follow-up PR. We waited for more than a week for the comments already.

**Comment by tcharding** [comment]
BOOM!

**Pull request merged**
